Replace mbed_mem_tracing_enabled macro with config option

kegilbert opened the pull request:
### Description

Move the memory tracing enabled macro to a config option - https://github.com/ARMmbed/mbed-os/issues/5225

There are a few macros that'd benefit from this (heap/stack/etc reporting for one). This does however break existing mbed_app.json files that relied on just defining the macro (you should get compile time warnings about redefinitions if you are defining the macro separately):

```
[Warning] mbed_config.h@147,0: "MBED_MEM_TRACING_ENABLED" redefined  
```
Is this a reasonable path forward? How should we handle transitioning to the config option definition rather than if the macro exists style?

**NOTE** Need to update the docs to match this as well. See - https://github.com/ARMmbed/mbed-os-5-docs/issues/275
### Pull request type

    [ ] Fix
    [X] Refactor
    [ ] New target
    [ ] Feature
    [X] Breaking change

kegilbert committed:
Replace mbed_mem_tracing_enabled macro with config option

geky reviewed (approved):
This looks reasonable to me

0xc0170 commented:
> Is this a reasonable path forward? How should we handle transitioning to the config option definition rather than if the macro exists style?

@ARMmbed/mbed-os-tools Can tools handle this - make changes like this non-breaking? 

@kegilbert The breaking change is that it would produce warnings and a user should remove the macro, and use a config? Or is  there anything else to be done?

theotherjimmy commented:
@0xc0170 There is no current method for deprecating ~config parameters~ macros and replacing them with config params.

@kegilbert Please avoid using internal links.

kegilbert commented:
@0xc0170 Correct, you should get some build warnings about the macro being redefined. The macro definition should be removed and replaced by the equivalent config option.

Updated the links as well.

kegilbert commented:
Have a PR up to fix the pr-head failure: https://github.com/ARMmbed/mbed-os-cliapp/pull/436

If this is passed review we can remove the `do no merge` tag and move forward with the patch PR above to move this one along.

0xc0170 reviewed (approved):
LGTM. Considering if used as it was, would lead to a warning that can be easily eliminated and proper config set.

0xc0170 commented:
> Have a PR up to fix the pr-head failure: ARMmbed/mbed-os-cliapp#436

It's still not merged, any progress?

kegilbert commented:
We have a bit of a chicken and egg issue with the CI, haven't heard back from Seppo on a recommended path forward (haven't pushed the issue to be fair, got caught up with other projects). The issue is I need the changes to go into Mbed OS so I can update the mbed_lib.json commit hash to include these changes in the mbed-cli repo so that can pass its CI. However I can't merge the changes here until the mbed-cli CI passes. 

What I could do is rebase my changes in a separate branch and update the mbed_lib hash in mbed-cli with that branch but seems a bit jank to have master mbed-cli pointing to a branch of mbed-os. Thoughts?

@0xc0170

0xc0170 commented:
@SeppoTakalo or @tommikas  would be better to advice here how to proceed.

cmonr commented:
@kegilbert Something you could do is update mbed-cli to this pr's head ref. I had to do something similar when updating mbed-cli for py3.

kegilbert commented:
@cmonr This no longer has a dependency on the mbed-os-cliapp changes as those don't get built in pr-head right now. This is passing Travis, should I remove the do not merge tag? For after the patch release code freeze.

cmonr commented:
@kegilbert Ah, didn't realize that was the reason DNM was added (I should've added a note).
Sure, you can remove it since that's the case. I was going to suggest that since it's a breaking change, it should be targeted for 5.10 anyways.

cmonr commented:
@kegilbert @ARMmbed/mbed-os-test Does anything in CI need to be updates and/or disabled to test this PR? Will any changes be needed after it's merged?

kegilbert commented:
@cmonr Only a docs update and patch note for the minor breaking change I believe, don't think this macro is used at the moment in Mbed OS tests outside of the one I changed in this PR:
https://github.com/ARMmbed/mbed-os/pull/7402/files#diff-9d0b4d826e1c32b0b1bfbe95983c8d8eR27

kegilbert commented:
@cmonr 
Handbook update: https://github.com/ARMmbed/mbed-os-5-docs/pull/684

cmonr commented:
/morph build
